### DISCUSSION TITLE
Differential Signals/Clocks

### DISCUSSION THREAD

**Discussion by davekeeshan** [cocotb/cocotb#3456] [Q&A]
Does any one have an example of a differential clock or a differential vector?

I am trying to encode a bus to do the dvi tdms protocol.  I want to have all the signals maintained inside cocotb (previous to this I was essentially using single ended signals and inverted the input on the second).

I would like to set up the clock so that it is treated like an ordinary clock except it controls 2 signals.

Similarly with data dignals, expecially if it is a vector.  I want to only set the postive half and the negative half tracks accordingly.  Bonus points if one can get to to work at the slice indexing level.

**Comment by ktbarrett** [comment]
Here's a mock up. It's not tested.

```python
class DifferentialClockDriver:

    def __init__(self, a, b, period, a_initial_value = 1):
        self.a = a
        self.b = b
        self.period = period
        self.a.value = a_initial_value
        self.b.value = int(not a_initial_value)
        self._task = None

    async def _run(self):
        while True:
            self.a.value = int(not self.a.value.integer)
            self.b.value = int(not self.a.value.integer)
            await Timer(self.period[0] / 2, self.period[1])

    def start(self):
        if self._task is not None:
            raise RuntimeError("clock is already started")
        self._task = cocotb.start_soon(self._run())

    def stop(self):
        if self._task is None:
            raise RuntimeError("clock has not been started")
        self._task.cancel()


diff_clock_driver = DifferentialClockDriver(cocotb.top.clkp, cocotb.top.clkn, period=(10, 'ns'), a_initial_value = 1)
diff_clock_driver.start()
diff_clock_driver.stop()
```

**Reply by cmarqu** [reply]
Complications could arise if you get a halfperiod that doesn't fit the simulator's timeprecision.
Potential solutions involve `decimals.Decimal` and `fractions.Fraction`, see
https://github.com/cocotb/cocotb/blob/52171eed34f6508bab0ad422b2f676ddcd0716f0/tests/test_cases/test_cocotb/test_clock.py#L68-L76

**Comment by davekeeshan** [comment]
This is what I did for the clock seemed the easier of the two, it inherits from the `Clock` Base Class:

Usage:
```python
    start_soon(DiffClock(self.clk_p, self.clk_n, self.clock_period, units="ns").start())        
````

```python
import itertools
from decimal import Decimal
from numbers import Real
from typing import Union
from cocotb.triggers import Timer
from cocotb.clock import Clock
  
class DiffClock(Clock):
    def __init__(
        self, signal_p , signal_n, period: Union[float, Real, Decimal], units: str = "step"
    ):
        Clock.__init__(self, signal_p, period, units)
        self.signal_n = signal_n


    async def start(self, cycles=None, start_high=True,wait_cycles=None):
        t = Timer(self.half_period)
        if cycles is None:
            it = itertools.count()
        else:
            it = range(cycles)

        for _ in it:
            self.signal.value   = start_high
            self.signal_n.value = not start_high
            await t
            self.signal.value   = not start_high
            self.signal_n.value = start_high
            await t
```

**Comment by davekeeshan** [comment]
It is the differential vector I get killed on.  The code below mostly work, except when one want to change a single bit inside the vector, that is where I get stuck

so `val[n] = 1` works , but is deprecated (might even have been removed this past week, that is the `__setitem__` function)
the official way, `val[n].value = 1` does not work, it only set one bit and it's mirror remains untouched.

Usage:
```python    
    self.data = DiffModifiableObject(self.data_p, self.data_n)
```

```python
from cocotb.handle import ModifiableObject

class DiffModifiableObject(ModifiableObject):
    
    def __init__(self, p, n):
        ModifiableObject.__init__(self, p._handle, p._path)
        self.n = ModifiableObject(n._handle, n._path)
        self.mask = (1 << len(self)) -1

    def __setitem__(self, index, value):
        ModifiableObject.__setitem__(self, index, value)
        self.n[index].value = not(bool(value))       
         
    def _set_value(self, value, call_sim):
        ModifiableObject._set_value(self, value, call_sim)
        self.n._set_value(value ^ self.mask, call_sim)
```

**Reply by ktbarrett** [reply]
This is an unfortunate limiataion of some simulator's VPI support. They support finding all the elements of a vector, but don't support settings values on them (at least I remember seeing this in Questa). The best thing to do is to make a separate driver and update the full vector at a time.

```python
self.vector.value = 0b01
self.vector.value = 0b10
```

**Reply by davekeeshan** [reply]
@ktbarrett 

Yes that is what the above code does actually, it subclasses `ModifiableObject`, so it uses its `value`, the `p` (`p` is implied here), as one half and then add another `ModifiableObject` and the `n`.  When you then use it in test code, you set the vector once but the two values get set.  It was setting a single bit of the vector I am stuck with.  VPI of course has it limits which is why up move up the stack and do it some where common beforew the final vectors are presented to the simulator
